### PR TITLE
util.timeutil, util.hlc:improve timeutil.Now() and hlc.Clock

### PR DESCRIPTION
**Pull request by andreimatei**
See individual commits.

This is meant to replace #71355 and #81139.

### CODE REVIEW

File: pkg/util/timeutil/time_test.go [hunk 1]
```diff
@@ -70,3 +72,43 @@ func TestReplaceLibPQTimePrefix(t *testing.T) {
...
+func TestTimeIsUTC(t *testing.T) {
+	require.Equal(t, time.UTC, Now().Location())
+	require.Equal(t, time.UTC, Unix(1, 1).Location())
+}
+
+// Test that the unsafe cast we do in timeutil.Now() to set the time to UTC
+// appears to be sane.
+func TestTimeUnsafeUTCCast(t *testing.T) {
+	require.Equal(t, unsafe.Sizeof(time.Time{}), unsafe.Sizeof(timeLayout{}))
+}
```
Comment (ajwerner) [inline pkg/util/timeutil/time_test.go:114]: you can use the `reflect` package to inspect the fields of the stdlib type, and I think you should. Consider this test: https://go.dev/play/p/tY0YRBv7YzL

File: pkg/util/timeutil/time_source.go [hunk 2]
```diff
@@ -54,9 +54,14 @@ type TickerI interface {
...
 }
 
-// DefaultTimeSource is a TimeSource using the system clock.
+// DefaultTimeSource is a TimeSource using the system clock. Use
+// SystemTimeSource for a singleton.
 type DefaultTimeSource struct{}
 
+// SystemTimeSource is an instance of DefaultTimeSource that can be used as a
+// singleton.
+var SystemTimeSource = DefaultTimeSource{}
```
Comment (ajwerner) [inline pkg/util/timeutil/time_source.go:63]: why? empty structs have a fun property that they always have the same address: https://go.dev/play/p/qe2WDZ-lXBI

File: pkg/util/hlc/hlc.go [hunk 3]
```diff
@@ -197,6 +229,28 @@ func (m *HybridManualClock) Resume() {
...
+func NewClockWithSystemTimeSource(maxOffset time.Duration) *Clock {
+	return NewClockWithTimeSource(timeutil.SystemTimeSource, maxOffset)
+}
+
+// NewClockWithTimeSource returns a Clock configured to use a specified time
+// source.
+//
+// A value of 0 for maxOffset means that clock skew checking, if performed on
+// this clock by RemoteClockMonitor, is disabled.
+func NewClockWithTimeSource(timeSource NowSource, maxOffset time.Duration) *Clock {
```
Comment (ajwerner) [inline pkg/util/hlc/hlc.go]: What's with the misalignment of `TimeSource` on the functions and `NowSource` as the type name?

**Review by ajwerner — COMMENTED** [review]
Should you rename `hlc.ManualClock` to something like `hlc.ManualTimeSource`?

**Review by andreimatei — COMMENTED** [review]
> Should you rename hlc.ManualClock to something like hlc.ManualTimeSource?

There's also `timeutil.ManualTime`, which is a `timeutil.TimeSource`. Do you think we should get rid of `hlc.ManualClock`, and replace all its uses with the more powerful `timeutil.ManualTime` ? If so, consider whether you'll allow me to do that in a separate PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-:-N31_iknA0v1JF1JGo8-:by2vchv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @nvanbenschoten)

___
*[`pkg/util/hlc/hlc.go` line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-N2O6uywholexnca5CG7-r2-247:-N318kECDtPgPCM--mhf:b-bpifzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/258347a8b1778f235583a9be6c22177b2273a2ba/pkg/util/hlc/hlc.go#L247)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What's with the misalignment of `TimeSource` on the functions and `NowSource` as the type name?
</blockquote></details>

it's not a beaut, but I'm not sure I prefer anything else. The interface is called `hlc.NowSource` instead of `hlc.TimeSource` because I didn't want it to clash with `timeutil.TimeSource`. Btw, the reason why I've introduced this new interface at all, instead of using `timeutil.TimeSource` (`NowSource` is a sub-interface of `TimeSource`) is so that `hlc.HybridManualClock` and others functors can implement `NowSource` without needing to deal with timers.
So, at the type name level, I want to use something different from `TimeSource`, but at the variable level I think `nowSource` is kind of a silly name. What would you suggest?
___
*[`pkg/util/timeutil/time_test.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-N2O6uyvHm4voo1qc2Z8-r1-114:-N31J8do7xEW7fR8FIZp:b-lhs69o) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6750fa224275ae705dd73f1190d229e04268e/pkg/util/timeutil/time_test.go#L114)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

you can use the `reflect` package to inspect the fields of the stdlib type, and I think you should. Consider this test: https://go.dev/play/p/tY0YRBv7YzL
</blockquote></details>

done, thanks
___
*[`pkg/util/timeutil/time_source.go` line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-N2O6uywholexnca5CG9-r2-63:-N31G243AQXYIZVDFhG8:blyk3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/258347a8b1778f235583a9be6c22177b2273a2ba/pkg/util/timeutil/time_source.go#L63)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why? empty structs have a fun property that they always have the same address: https://go.dev/play/p/qe2WDZ-lXBI
</blockquote></details>

ok, I've removed it. Although I personally still like singletons - because the user doesn't need to understand that the struct is empty.

**Review by andreimatei — COMMENTED** [review]
I've had to introduce a `timeutil.NowNoMono()` in the first commit, to keep some tests that wanted `time.Time` instances to roundtrip through yaml and crap working. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-:-N31aEjoATNLSi_hzi8s:bx4onef)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @nvanbenschoten)

**Review by ajwerner — APPROVED** [review]
You've still got some failing tests. The remaining chatter is about names. I think we should go and rename a bunch of these so the concepts line up, but you can leave that for follow-up. The logic of the change :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-:-N323rCOCg5c3Jb6M1sO:btwezmy)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @dt, and @nvanbenschoten)

___
*[`pkg/util/hlc/hlc.go` line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-N2O6uywholexnca5CG7-r2-247:-N322ReJ2F03Wk--lHkW:bs3n7tt) ([raw file](https://github.com/cockroachdb/cockroach/blob/258347a8b1778f235583a9be6c22177b2273a2ba/pkg/util/hlc/hlc.go#L247)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it's not a beaut, but I'm not sure I prefer anything else. The interface is called `hlc.NowSource` instead of `hlc.TimeSource` because I didn't want it to clash with `timeutil.TimeSource`. Btw, the reason why I've introduced this new interface at all, instead of using `timeutil.TimeSource` (`NowSource` is a sub-interface of `TimeSource`) is so that `hlc.HybridManualClock` and others functors can implement `NowSource` without needing to deal with timers.
So, at the type name level, I want to use something different from `TimeSource`, but at the variable level I think `nowSource` is kind of a silly name. What would you suggest?
</blockquote></details>

What if we rename `hlc.NowSource` to `WallClock` or `NanoTimeClock` or `UnixNanoClock`? 


 I'll be honest, I don't love `TimeSource` and I regret it as a name. How crazy would it be to rename `timeutil.TimeSource` to `timeutil.Timeline`? I don't know if that's better. I was thinking it's better because the timer events happen on the `Timeline`? We can leave renaming for future PRs I suppose.

**Review by andreimatei — COMMENTED** [review]
> Do you think we should get rid of hlc.ManualClock, and replace all its uses with the more powerful timeutil.ManualTime ?

Did you have an opinion on this? Cause I'm tempted to do it, in a separate PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-:-N35FYsP0bHB9SvCI6a0:b-7w1by2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, and @nvanbenschoten)

___
*[`pkg/util/hlc/hlc.go` line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81484#-N2O6uywholexnca5CG7-r2-247:-N35E_md9k_c3xDuvvmv:bcxd0l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/258347a8b1778f235583a9be6c22177b2273a2ba/pkg/util/hlc/hlc.go#L247)):*
> What if we rename hlc.NowSource to WallClock or NanoTimeClock or UnixNanoClock?

Went with `WallClock`.

> I'll be honest, I don't love TimeSource and I regret it as a name. How crazy would it be to rename timeutil.TimeSource to timeutil.Timeline? I don't know if that's better. I was thinking it's better because the timer events happen on the Timeline? We can leave renaming for future PRs I suppose.

I don't think I line `Timeline`... It's not suggestive enough. I think `TimeSource` is good; it's kinda synonymous with `Clock`, but perhaps suggests that there's a bit more going on - the timer stuff.

**Comment by andreimatei** [comment]
TFTR

bors r+

**Pull request merged**